src/CachePlugin.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace Http\Client\Plugin;
+
+use Psr\Cache\CacheItemPoolInterface;
+use Psr\Http\Message\RequestInterface;
+use Psr\Http\Message\ResponseInterface;
+
+/**
+ * Allow for caching a response.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+class CachePlugin implements Plugin
+{
+    /**
+     * @var CacheItemPoolInterface
+     */
+    private $pool;
+
+    /**
+     * Default time to store object in cache. This value is used if CachePlugin::respectCacheHeaders is false or
+     * if cache headers are missing.
+     *
+     * @var int
+     */
+    private $defaultTtl;
+
+    /**
+     * Look at the cache headers to know how long this response is going to be cached.
+     *
+     * @var bool
+     */
+    private $respectCacheHeaders;
+
+    /**
+     * @param CacheItemPoolInterface $pool
+     * @param array                  $options
+     */
+    public function __construct(CacheItemPoolInterface $pool, array $options = array())
+    {
+        $this->pool = $pool;
+        $this->defaultTtl = isset($options['default_ttl']) ? $options['default_ttl'] : null;
+        $this->respectCacheHeaders = isset($options['respect_cache_headers']) ? $options['respect_cache_headers'] : true;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function handleRequest(RequestInterface $request, callable $next, callable $first)
+    {
+        $method = strtoupper($request->getMethod());
+
+        // if the request not is cachable, move to $next
+        if ($method !== 'GET' && $method !== 'HEAD') {
+            return $next($request);
+        }
+
+        // If we can cache the request
+        $key = $this->createCacheKey($request);
+        $cacheItem = $this->pool->getItem($key);
+
+        if ($cacheItem->isHit()) {
+            // return cached response
+            return $cacheItem->get();
+        }
+
+        return $next($request)->then(function (ResponseInterface $response) use ($cacheItem) {
+            if ($this->isCacheable($response)) {
+                $cacheItem->set($response)
+                    ->expiresAfter($this->getMaxAge($response));
+                $this->pool->save($cacheItem);
+            }
+
+            return $response;
+        });
+    }
+
+    /**
+     * Verify that we can cache this response.
+     *
+     * @param ResponseInterface $response
+     *
+     * @return bool
+     */
+    private function isCacheable(ResponseInterface $response)
```

Comment:
Can we have this protected ? So if someone want to have its own rule for caching a reponse it is possible without rewriting everything.